cmd/dep/root_analyzer.go

```diff
@@ -53,6 +63,65 @@ func (a *rootAnalyzer) InitializeRootManifestAndLock(dir string, pr gps.ProjectR
 	return
 }
 
+func (a *rootAnalyzer) cacheDeps(pr gps.ProjectRoot) error {
+	deps := make(map[gps.ProjectRoot]bool)
+	logger := a.ctx.Err
+	g, ctx := errgroup.WithContext(context.TODO())
+	sem := make(chan struct{}, concurrency)
+
+	syncDep := func(pr gps.ProjectRoot, sm gps.SourceManager) error {
+		if err := sm.SyncSourceFor(gps.ProjectIdentifier{ProjectRoot: pr}); err != nil {
+			logger.Printf("Unable to cache %s - %s", pr, err)
+			return err
+		}
+		logger.Printf("Cached %s", pr)
+		return nil
+	}
+
+	for ip := range a.directDeps {
+		logger.Printf("Package %q, analyzing...", ip)
+		if paths.IsStandardImportPath(ip) {
+			continue
+		}
+		if hasImportPathPrefix(ip, string(pr)) {
+			continue
+		}
+
+		pr, err := a.sm.DeduceProjectRoot(ip)
+		if err != nil {
+			return err
+		}
+
+		if _, ok := deps[pr]; ok {
+			continue
+		}
+
+		g.Go(func() error {
+			select {
+			case sem <- struct{}{}:
```

Comment:
I realize now that the existing usage of using errgroup in dep was not the best/canonical example to have pointed you to.

I would rather not use this extra sem channel to handle bounded parallelism and instead use the pattern from the errgroup documentation: 

https://godoc.org/golang.org/x/sync/errgroup#example-Group--Pipeline

In that example, they start a fixed number of goroutines instead of a goroutine per work item (look for `numDigesters`), and range over a channel of input values (in our case that would be the project roots).